src/topology/builder.rs

```diff
@@ -337,12 +338,80 @@ pub async fn build_pieces(
                 // of relying on the one fallible function pattern we currently have
                 assert_eq!(1, named_outputs.len());
                 outputs.insert(
-                    OutputId::from((key, named_outputs.remove(0))),
+                    OutputId::from((key, named_outputs.remove(0).0)),
                     errors_control,
                 );
 
                 Task::new(key.clone(), typetag, transform)
             }
+            Transform::DispatchFunction(mut t) => {
+                let mut output_fans = IndexMap::new();
+                let mut output_controls = IndexMap::new();
+                let output_names = named_outputs.clone();
+                for (name, _) in output_names.iter() {
+                    let (output, control) = Fanout::new();
+                    output_fans.insert(name.clone(), output);
+                    output_controls.insert(name.clone(), control);
+                }
+
+                let mut input_rx = input_rx
+                    .filter(move |event| ready(filter_event_type(event, input_type)))
+                    .ready_chunks(128); // 128 is an arbitrary, smallish constant
+
+                let mut timer = crate::utilization::Timer::new();
+                let mut last_report = Instant::now();
+
+                let transform = async move {
+                    timer.start_wait();
+                    while let Some(events) = input_rx.next().await {
+                        let stopped = timer.stop_wait();
+                        if stopped.duration_since(last_report).as_secs() >= 5 {
+                            timer.report();
+                            last_report = stopped;
+                        }
+
+                        emit!(&EventsReceived {
+                            count: events.len(),
+                            byte_size: events.size_of(),
+                        });
+
+                        // TODO: optimize the creation of buffers
+                        let mut output_bufs = Vec::with_capacity(output_names.len() * events.len());
+                        let mut bufs = Vec::with_capacity(output_names.len());
```

Comment:
Two things:
- can we collapse this to a single output buffer that gets used over and over instead of a temporary buffer that we just append to the "main" buffer?
- I think the default capacity should simply be `events.len()`

On point #1, my assumption is that transforms are not actually clearing the vector before pushing into it, so we can save ourselves some copies/allocations by using one single buffer.

On point #2, my assumption is that for these multi-output transforms, the majority of them are working as a router, so single output, not fanout.  We'd be wasting a lot of pre-allocated capacity by using a default that assumes we're going to fan the event out to every single named output in the transform.

Reply:
As a follow-up comment to point #1, maybe instead of passing a raw reference to `&mut Vec<...>`, we should actually change the trait signature to take `output: O` where `O: Extend<Event>` so that we can simultaneously switch to using a single output buffer, and enforce the concept that the buffer is append only.